src/infer_payload.cc

```diff
@@ -38,13 +38,15 @@ InferPayload::InferPayload(
 
 InferPayload::~InferPayload()
 {
+  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
question: My understanding is that the std::uniqueptr<promise> will be destroyed through the normal destructor - and there's no need for a lock or an explicit reset here